# Migrating to Amazon Bedrock from OpenAI or Anthropic

> **Sample code — not for production.** Provided as AWS Content under the AWS
> Customer Agreement; do not use it in production accounts or on production or other
> critical data. Running these cells calls Amazon Bedrock and incurs charges. Full
> disclaimer in the [README](../README.md#disclaimer).

Code written for the OpenAI or Anthropic SDK moves to Amazon Bedrock by changing the
base URL, the key and the model ID. This notebook makes that change for both SDKs on
both endpoints, then handles what a straight port gets wrong: model IDs, URL paths,
sampling parameters and the API each model serves, with a small shim that lets one call
site work across families.

**Covered:** the OpenAI and Anthropic SDKs on both endpoints, model IDs, one client per
URL path, a compatibility shim, streaming on the three APIs, LLM gateways, what you gain
and lose, and a checklist.

## Setup

Install the pinned requirements first (`pip install -r ../requirements.txt`, Python 3.11
or later). Both SDKs authenticate with a short-term Bedrock API key minted from your IAM
credentials;
[`00-foundations/01`](../00-foundations/01-endpoints-auth-and-the-three-paths.ipynb)
explains the key, the endpoints and the paths. From `_shared/bedrock.py`, this
repository's helper module, the notebook takes `api_prefix`, which picks a model's URL
path on each endpoint.

In [1]:
import re
import sys

import anthropic
import openai
from aws_bedrock_token_generator import provide_token
from openai import OpenAI

sys.path.insert(0, "../_shared")
from bedrock import api_prefix

REGION = "us-east-1"
MANTLE = f"https://bedrock-mantle.{REGION}.api.aws"
RUNTIME = f"https://bedrock-runtime.{REGION}.amazonaws.com"
KEY = provide_token(region=REGION)  # lasts at most 12 hours; mint a new one before then
QUESTION = "What is the capital of Australia? Answer in one word."

## 1. The OpenAI SDK

For OpenAI code the change is `base_url`, `api_key` and the model ID. `bedrock-runtime`,
which AWS recommends for new applications, serves the OpenAI APIs at `/openai/v1`, and
several families there need an inference profile such as `us.openai.gpt-5.6-sol`.
Sending the `bedrock-mantle` ID `openai.gpt-5.6-sol` to `bedrock-runtime` returns
`Invocation of model ID openai.gpt-5.6-sol with on-demand throughput isn’t supported`,
and an ID that differs between the endpoints returns `The provided model identifier is
invalid`. Choose `bedrock-mantle` when you need something only it has (section 8).

In [2]:
# Before: client = OpenAI(api_key=os.environ["OPENAI_API_KEY"])
DESTINATIONS = [
    ("bedrock-runtime", RUNTIME + "/openai/v1", "us.openai.gpt-5.6-sol"),
    ("bedrock-mantle", MANTLE + "/openai/v1", "openai.gpt-5.6-sol"),
]
for label, base_url, model in DESTINATIONS:
    client = OpenAI(base_url=base_url, api_key=KEY, timeout=180, max_retries=5)
    response = client.responses.create(model=model, input=QUESTION, max_output_tokens=2000, store=False)
    print(f"{label:16} {model:24} -> {response.output_text.strip()}")

bedrock-runtime  us.openai.gpt-5.6-sol    -> Canberra


bedrock-mantle   openai.gpt-5.6-sol       -> Canberra


The same change through environment variables, for code you would rather not edit. The
model ID still has to change, because the variables do not translate it.

```bash
export OPENAI_BASE_URL="https://bedrock-runtime.us-east-1.amazonaws.com/openai/v1"
export OPENAI_API_KEY="$(python -c 'from aws_bedrock_token_generator import provide_token; print(provide_token(region="us-east-1"))')"
```

Do not hardcode a long-term key. A short-term key is a presigned SigV4 request that
carries the minting role's permissions and expires within 12 hours.

## 2. The Anthropic SDK

For Anthropic code, point `base_url` at the `/anthropic` path of either endpoint,
without `/v1`, which the SDK adds. Pass the Bedrock key as `auth_token`, which the SDK
sends as a bearer token, with `api_key=None`. On `bedrock-runtime` Claude needs an
inference profile such as `us.anthropic.claude-opus-5`; the reply's `model` field
carries the short model name either way.

In [3]:
for label, base_url, model in (
    ("bedrock-runtime", RUNTIME + "/anthropic", "us.anthropic.claude-opus-5"),
    ("bedrock-mantle", MANTLE + "/anthropic", "anthropic.claude-opus-5"),
):
    claude = anthropic.Anthropic(base_url=base_url, auth_token=KEY, api_key=None, timeout=180, max_retries=5)
    message = claude.messages.create(model=model, max_tokens=200, messages=[{"role": "user", "content": QUESTION}])
    text = "".join(block.text for block in message.content if block.type == "text")
    print(f"{label:16} {model:28} -> {text.strip()}   (model field: {message.model})")

bedrock-runtime  us.anthropic.claude-opus-5   -> Canberra   (model field: claude-opus-5)


bedrock-mantle   anthropic.claude-opus-5      -> Canberra   (model field: claude-opus-5)


## 3. Model IDs

OpenAI and Anthropic model names do not exist on Bedrock. Map each one to a Bedrock
model before moving traffic. The mapping below is a starting point by rough capability,
so evaluate each replacement on your own tasks. The cell checks every target against
this Region's `bedrock-mantle` catalogue and shows what an OpenAI name returns.

In [4]:
MODEL_MAP = {
    "gpt-4o": "openai.gpt-5.6-sol",
    "gpt-4o-mini": "openai.gpt-5.4",
    "o1": "openai.gpt-5.6-sol",
    "gpt-3.5-turbo": "openai.gpt-oss-20b",
    "claude-3-5-sonnet": "anthropic.claude-sonnet-5",
    "claude-3-haiku": "anthropic.claude-haiku-4-5",
}
catalogue = {m.id for m in OpenAI(base_url=MANTLE + "/v1", api_key=KEY, timeout=60).models.list()}
for old, new in MODEL_MAP.items():
    print(f"{old:18} -> {new:28} in the catalogue: {new in catalogue}")

try:
    OpenAI(base_url=MANTLE + "/openai/v1", api_key=KEY, timeout=60, max_retries=0).responses.create(
        model="gpt-4o", input=QUESTION, max_output_tokens=16, store=False)
except openai.NotFoundError as exc:
    print(f"\ngpt-4o -> {exc.status_code} {exc.body.get('message') if isinstance(exc.body, dict) else exc.body}")

gpt-4o             -> openai.gpt-5.6-sol           in the catalogue: True
gpt-4o-mini        -> openai.gpt-5.4               in the catalogue: True
o1                 -> openai.gpt-5.6-sol           in the catalogue: True
gpt-3.5-turbo      -> openai.gpt-oss-20b           in the catalogue: True
claude-3-5-sonnet  -> anthropic.claude-sonnet-5    in the catalogue: True
claude-3-haiku     -> anthropic.claude-haiku-4-5   in the catalogue: True



gpt-4o -> 404 The model 'gpt-4o' does not exist


## 4. One client per URL path

One base URL does not cover every family. On `bedrock-mantle` the hosted GPT models,
Gemma 4 and Grok are at `/openai/v1`, gpt-oss and the other open-weight families at
`/v1`, and Claude at `/anthropic/v1`; on `bedrock-runtime` every OpenAI-compatible model
is at `/openai/v1`. A model sent to the wrong path returns 400 ``model
`openai.gpt-oss-120b` isn't supported on this route``. `api_prefix` picks the path, and
the factory below keeps one client per base URL.

In [5]:
_clients: dict[str, OpenAI] = {}


def client_for(model: str, endpoint: str = "mantle") -> OpenAI:
    """An OpenAI SDK client on the URL path this model needs."""
    base_url = (MANTLE if endpoint == "mantle" else RUNTIME) + api_prefix(model, endpoint)
    if base_url not in _clients:
        _clients[base_url] = OpenAI(base_url=base_url, api_key=KEY, timeout=180, max_retries=5)
    return _clients[base_url]


for model in ("openai.gpt-5.6-sol", "openai.gpt-oss-120b", "google.gemma-4-31b", "qwen.qwen3-32b"):
    client = client_for(model)
    reply = client.chat.completions.create(model=model, messages=[{"role": "user", "content": QUESTION}],
                                           max_completion_tokens=2000)
    print(f"{model:22} {str(client.base_url):48} -> {(reply.choices[0].message.content or '').strip()}")

openai.gpt-5.6-sol     https://bedrock-mantle.us-east-1.api.aws/openai/v1/ -> Canberra


openai.gpt-oss-120b    https://bedrock-mantle.us-east-1.api.aws/v1/     -> Canberra


google.gemma-4-31b     https://bedrock-mantle.us-east-1.api.aws/openai/v1/ -> Canberra


qwen.qwen3-32b         https://bedrock-mantle.us-east-1.api.aws/v1/     -> Canberra


## 5. A compatibility shim

Sampling parameters and service tiers are accepted per model, so a shim that carries a
table of restrictions goes stale. The service names the parameter it refuses, as in
`Unsupported parameter: 'top_p' is not supported with this model.` or `` `temperature`
is deprecated for this model.``, so the shim below sends only the options you pass, and
when a 400 names one of them, drops it and tries again, once per option, saying what it
dropped. It uses Messages for Claude and Chat Completions for every other model, because
every other family on `bedrock-mantle` serves Chat Completions and accepts
`max_completion_tokens`, as
[`01-choosing-a-model-and-api`](01-choosing-a-model-and-api.ipynb) measures.

In [6]:
TUNABLE = ("temperature", "top_p", "service_tier")
CLAUDE = anthropic.Anthropic(base_url=MANTLE + "/anthropic", auth_token=KEY, api_key=None,
                             timeout=180, max_retries=5)


def refused_option(exc: Exception) -> str | None:
    """The option in TUNABLE that a 400 names, if it names one."""
    body = exc.body if isinstance(getattr(exc, "body", None), dict) else {}
    error = body.get("error")
    message = str((error.get("message") if isinstance(error, dict) else None) or body.get("message") or "")
    quoted = [name for name in re.findall(r"[`'\"]([a-z_]+)[`'\"]", message) if name in TUNABLE]
    return quoted[0] if quoted else next((name for name in TUNABLE if name in message), None)


def send(model: str, prompt: str, max_tokens: int, options: dict) -> str:
    """One request: Messages for Claude, Chat Completions for every other model."""
    messages = [{"role": "user", "content": prompt}]
    if model.startswith("anthropic."):
        reply = CLAUDE.messages.create(model=model, max_tokens=max_tokens, messages=messages, **options)
        return "".join(block.text for block in reply.content if block.type == "text")
    reply = client_for(model).chat.completions.create(model=model, messages=messages,
                                                      max_completion_tokens=max_tokens, **options)
    return reply.choices[0].message.content or ""

In [7]:
def complete(model: str, prompt: str, *, max_tokens: int = 2000, **options) -> str:
    """Send the options you pass; drop each one a 400 names, once, and say so."""
    dropped = []
    for _ in range(len(options) + 1):
        try:
            text = send(model, prompt, max_tokens, options)
        except (openai.BadRequestError, anthropic.BadRequestError) as exc:
            name = refused_option(exc)
            if name not in options:
                raise
            options.pop(name)
            dropped.append(name)
            continue
        if dropped:
            print(f"  [{model}: dropped {', '.join(dropped)}]")
        return text
    raise RuntimeError(f"{model}: still refused after dropping {dropped}")

In [8]:
# The same call site for five models, each passed options some of them refuse.
for model in ("openai.gpt-5.6-sol", "google.gemma-4-31b", "qwen.qwen3-32b",
              "anthropic.claude-haiku-4-5", "anthropic.claude-sonnet-5"):
    answer = complete(model, QUESTION, temperature=0.7, top_p=0.95, service_tier="flex")
    print(f"{model:28} -> {answer.strip()}")

  [openai.gpt-5.6-sol: dropped service_tier, temperature, top_p]
openai.gpt-5.6-sol           -> Canberra


google.gemma-4-31b           -> Canberra
qwen.qwen3-32b               -> Canberra


  [anthropic.claude-haiku-4-5: dropped service_tier, temperature]
anthropic.claude-haiku-4-5   -> Canberra


  [anthropic.claude-sonnet-5: dropped service_tier, temperature, top_p]
anthropic.claude-sonnet-5    -> **Canberra**


## 6. Streaming on the three APIs

Each API streams a different event shape. The Responses API sends typed events with text
in `response.output_text.delta`, Chat Completions sends chunks with text in
`choices[0].delta.content`, and Messages sends `content_block_delta` events, which the
Anthropic SDK gathers into `text_stream`. The generator below yields plain text from any
of them. How many deltas arrive is up to the model: a short answer can come as one
delta, so the cell prints the count next to the assembled text.

In [9]:
def stream_text(model: str, prompt: str, api: str, max_tokens: int = 300):
    """Yield text deltas from the Responses API, Chat Completions or Messages."""
    if api == "messages":
        with CLAUDE.messages.stream(model=model, max_tokens=max_tokens,
                                    messages=[{"role": "user", "content": prompt}]) as stream:
            yield from stream.text_stream
    elif api == "responses":
        with client_for(model).responses.stream(model=model, input=prompt, max_output_tokens=max_tokens, store=False) as stream:
            for event in stream:
                if event.type == "response.output_text.delta":
                    yield event.delta
    else:
        for chunk in client_for(model).chat.completions.create(
                model=model, messages=[{"role": "user", "content": prompt}], max_completion_tokens=max_tokens,
                stream=True):
            if chunk.choices and chunk.choices[0].delta.content:
                yield chunk.choices[0].delta.content


for model, api in (("google.gemma-4-31b", "responses"), ("qwen.qwen3-32b", "chat"),
                   ("anthropic.claude-haiku-4-5", "messages")):
    pieces = list(stream_text(model, "Count from 1 to 10, separated by commas.", api))
    print(f"{model:28} {api:10} {len(pieces):3} deltas -> {''.join(pieces).strip()}")

google.gemma-4-31b           responses   29 deltas -> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10


qwen.qwen3-32b               chat         1 deltas -> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10


anthropic.claude-haiku-4-5   messages     4 deltas -> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10


## 7. LLM gateways

Gateways such as LiteLLM reach Bedrock through the same OpenAI-compatible URLs. Give
each model its own `api_base`, because the path depends on the family:

```yaml
model_list:
  - model_name: frontier
    litellm_params:
      model: openai/us.openai.gpt-5.6-sol
      api_base: https://bedrock-runtime.us-east-1.amazonaws.com/openai/v1
      api_key: os.environ/BEDROCK_API_KEY
  - model_name: open-weight
    litellm_params:
      model: openai/qwen.qwen3-32b
      api_base: https://bedrock-mantle.us-east-1.api.aws/v1
      api_key: os.environ/BEDROCK_API_KEY
```

The key expires within 12 hours and cannot be renewed, so mint a new one on a schedule,
well inside that lifetime, and hand it to the gateway in whatever way your deployment
reloads its environment:

```bash
export BEDROCK_API_KEY="$(python -c 'from aws_bedrock_token_generator import provide_token; print(provide_token(region="us-east-1"))')"
```

## 8. What you gain and what you lose

As of September 2026:

| Moving from | You gain | You lose or must change |
|---|---|---|
| The OpenAI API | IAM authentication, Projects, data retention controls, service tiers, models from several providers, and web search on the hosted GPT models | OpenAI model IDs, sampling parameters on some models, and `max_tokens` on GPT-5.6's Chat Completions |
| The Anthropic API | IAM authentication, Workspaces, and the same Claude models on `bedrock-runtime` through Converse | `output_config.format` on `bedrock-mantle` (use a forced tool), and `temperature` and `top_p` on newer Claude models, which refuse them as deprecated |
| `bedrock-runtime`, to `bedrock-mantle` | The bare `/v1` path, models only there such as Gemma 4 and Grok 4.3, bare model IDs with no profile, server-side tools such as web search, `background=true`, and the control plane for Projects, data retention, files and fine-tuning | Converse, cross-Region inference profiles, Provisioned Throughput, batch inference, and Guardrails, whose header every `bedrock-mantle` API ignores ([`03-production-hardening`](03-production-hardening.ipynb)) |

## 9. A migration checklist

1. Map each model name to a Bedrock model ID, per endpoint.
2. Point each model at its URL path; one base URL does not cover every family.
3. Remove shared sampling defaults, and add `temperature` or `top_p` back per model
   after testing.
4. Send `max_completion_tokens` on Chat Completions.
5. Check that each model serves the API your code calls.
6. Send `service_tier` only where the model accepts it.
7. Set `store` explicitly on the Responses API, which keeps responses for 30 days by
   default.
8. Replace static API keys with short-term keys minted from IAM, renewed before they
   expire.
9. Retry 429 and 5xx with backoff, and fail fast on other 4xx.
10. Set a client timeout on every call.
11. Check the values in structured output.
12. Point dashboards for `bedrock-mantle` traffic at the `AWS/BedrockMantle` namespace.
13. Give each workload a tagged Project for cost attribution.
14. Confirm that every model you depend on is in your Region.

## Things that differ

What trips people up here, and what to do about it.

| Behaviour | What you see | What to do |
|---|---|---|
| OpenAI model names | 404 `The model 'gpt-4o' does not exist` | Map each name to a Bedrock model ID (section 3). |
| The wrong URL path | 400 ``model `…` isn't supported on this route`` | Keep one client per path (section 4). |
| `max_tokens` on GPT-5.6 | 400 `Unsupported parameter: 'max_tokens' is not supported with this model.` | Send `max_completion_tokens` on Chat Completions. |
| Shared sampling defaults | A 400 that names the parameter | Send none by default, or drop what the 400 names (section 5). |
| A model without the Responses API | 400 `The model '…' does not support the '/v1/responses' API` | Use Chat Completions, which every family except Claude serves. |
| A `bedrock-mantle` ID on `bedrock-runtime` | `Invocation of model ID … with on-demand throughput isn’t supported`, or `The provided model identifier is invalid` | Use the runtime ID or its inference profile ([`00-foundations/01`](../00-foundations/01-endpoints-auth-and-the-three-paths.ipynb)). |

## Next

- Which model and API, measured across families:
  [`01-choosing-a-model-and-api`](01-choosing-a-model-and-api.ipynb)
- Before production traffic: [`03-production-hardening`](03-production-hardening.ipynb)
- Projects and data retention:
  [`00-foundations/02`](../00-foundations/02-governance-projects-and-retention.ipynb)